Fix new docker config format for private registries

tamnd opened the pull request:
Fix #12626

tamnd committed:
Support new docker config format for private registries

ngbinh commented on pkg/credentialprovider/config.go:
```diff
@@ -48,7 +55,8 @@ var (
 	rootDirPath       = "/"
 
-	configFileName = ".dockercfg"
+	configFileName     = ".dockercfg"
```
location has been moved https://github.com/docker/docker/blob/1e2765dabbc37979639b72d460e39a00422e77fe/cliconfig/config.go#L19

tamnd replied:
Fixed location:
```go
authHomeDirPath    = filepath.Join(homeDirPath, ".docker")
authRootDirPath    = filepath.Join(rootDirPath, ".docker")
```

ngbinh commented:
would be very nice if you can also fix the document https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/images.md

deads2k commented on pkg/credentialprovider/config.go:
```diff
@@ -64,6 +74,32 @@ func GetPreferredDockercfgPath() string {
 }
 
 func ReadDockerConfigFile() (cfg DockerConfig, err error) {
+	// Try happy path first - latest config file
```
How is docker handling the transition?  Are the files complementary?  If so, we should probably have two different provider instances to build the keyring.

tamnd replied:
[https://github.com/docker/docker/blob/master/cliconfig/config.go#L159](https://github.com/docker/docker/blob/master/cliconfig/config.go#L159)
In `func Load(configDir string) (*ConfigFile, error)`, first, they try to load new configuration file at `~/.docker/config.json` and parse by `configFile.LoadFromReader(file)`, if the config file is not found, then they fallback to load the old one at `~/.dockercfg` by `configFile.LegacyLoadFromReader(file)`.

deads2k replied:
> https://github.com/docker/docker/blob/master/cliconfig/config.go#L159
> In func Load(configDir string) (*ConfigFile, error), first, they try to load new configuration file at ~/.docker/config.json and parse by configFile.LoadFromReader(file), if the config file is not found, then they fallback to load the old one at ~/.dockercfg by configFile.LegacyLoadFromReader(file).

Can you add a comment explaining the behavior and why we're doing it?  It struck me as a little odd when I read it.

tamnd replied:
This function `func ReadDockerConfigFile() (cfg DockerConfig, err error)` is used to parse config which is stored in:
- `GetPreferredDockercfgPath() + "/config.json"`
- `workingDirPath + "/config.json"
- `$HOME/.docker/config.json`
- `/.docker/config.json`
- `GetPreferredDockercfgPath() + "/.dockercfg"`
- `workingDirPath + "/.dockercfg"
- `$HOME/.dockercfg`
- `/.dockercfg`

The first four one are new type of secret, and the last four one are the old type.

mattmoor replied:
I'd agree that having multiple providers, one that passes ".docker/config.json" and one that passes ".dockercfg" would be preferable because it blends the keyrings.  This is what I did here: https://github.com/kubernetes/kubernetes/pull/13894

However, I didn't realize this existed when I wrote that.

ngbinh replied:
I think having two separate providers for .docker/config.json" and ".dockercfg" makes sense. It clearly states in the issue that the format is also changed between the two so we cannot use the same secret for both providers.

mattmoor replied:
The new format is a strict superset of the old, so these are not mutually exclusive options.

The new format adds a level of JSON around an identical structure, and at the top-level can now also include `HttpHeaders`, which you aren't using yet.

This is clearly possible via `credentialprovider`, as I linked you a PR that does precisely this (ignoring `HttpHeaders`).

To support `HttpHeaders`, the `credentialprovider` interface will have to change anyways.

deads2k commented:
We also support having `dockercfg` files inside of `Secret` object stored via the REST API.  Those secrets can then be referenced inside of a `PodSpec` as an `imagePullSecret`.  I wouldn't hold up this pull for it, but you could also try updating that path down in: https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L282.

tamnd commented:
@deads2k I fixed the function `MakeDockerKeyring()` in [ttps://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L282](ttps://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/keyring.go#L282), which now can unmarshal new docker config format as you suggested.

I also added new `api.SecretType`: `SecretTypeDockerConfigJson` which contains a dockercfg file that follows the same format rules as `~/.docker/config.json` and `Secret.Data[".dockerconfigjson"]` contains  a serialized `~/.docker/config.json` file.

Here is how to specify image pull secrets with new docker config format:
[http://kubernetes.io/v1.0/docs/user-guide/images.html#specifying-imagepullsecrets-on-a-pod](http://kubernetes.io/v1.0/docs/user-guide/images.html#specifying-imagepullsecrets-on-a-pod)
```
$ docker login
Username: janedoe
Password: ●●●●●●●●●●●
Email: jdoe@example.com
WARNING: login credentials saved in /Users/jdoe/.docker/config.json.
Login Succeeded

$ echo $(cat ~/.docker/config.json)
{"auths": { "https://index.docker.io/v1/": { "auth": "ZmFrZXBhc3N3b3JkMTIK", "email": "jdoe@example.com" } }}

$ cat ~/.docker/config.json | base64
eyAiaHR0cHM6Ly9pbmRleC5kb2NrZXIuaW8vdjEvIjogeyAiYXV0aCI6ICJabUZyWlhCaGMzTjNiM0prTVRJSyIsICJlbWFpbCI6ICJqZG9lQGV4YW1wbGUuY29tIiB9IH0K

$ cat > /tmp/image-pull-secret.yaml <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: myregistrykey
data:
  .dockerconfigjson: eyAiaHR0cHM6Ly9pbmRleC5kb2NrZXIuaW8vdjEvIjogeyAiYXV0aCI6ICJabUZyWlhCaGMzTjNiM0prTVRJSyIsICJlbWFpbCI6ICJqZG9lQGV4YW1wbGUuY29tIiB9IH0K
type: kubernetes.io/dockerconfigjson
EOF

$ kubectl create -f /tmp/image-pull-secret.yaml
secrets/myregistrykey
```

What do you think about this approach? I will add more tests in `pkg/api/validation/validation_test.go` later.

mbforbes commented:
@k8s-bot ok to test

deads2k commented on pkg/credentialprovider/config.go:
```diff
@@ -30,6 +30,13 @@ import (
+// see https://github.com/docker/docker/pull/12009
+type DockerConfigJson struct {
+	Auths       DockerConfig      `json:"auths"`
+	HttpHeaders map[string]string `json:"HttpHeaders,omitempty"`
```
Can you add a comment about what this is for?  (Based on the docs here: https://docs.docker.com/reference/commandline/cli/, it looks like a set of blindly attached headers) and an indication that we don't use or respect them (yet).

@erictune Are you willing to take the `HttpHeaders` support as a follow on item?  I think we'd gain a lot of benefit even without plumbing them through.

tamnd replied:
This struct `DockerConfigJson` is used to store content of `~/.docker/config.json`, it is the same structure as [ConfigFile](https://github.com/docker/docker/blob/master/cliconfig/config.go#L57) struct in `docker/cliconfig`.
```go
// ConfigFile ~/.docker/config.json file info
type ConfigFile struct {
	AuthConfigs map[string]AuthConfig `json:"auths"`
	HTTPHeaders map[string]string     `json:"HttpHeaders,omitempty"`
	PsFormat    string                `json:"psFormat,omitempty"`
	filename    string                // Note: not serialized - for internal use only
}
```
In the case we don't use `HTTPHeaders` and `PsFormat`, then we do not need that struct. And as in your comment below, a single method `func DockerConfigFrom(bytes []bytes) (DockerConfig, error)` which parse both new and old type of secrets will make code easier to maintain.

erictune replied:
@deads2k Can you explain what you mean by "...take the HttpHeaders support as a follow on item... without plumbing them through.".

mattmoor replied:
I'd drop HttpHeaders until they are plumbed through.

deads2k replied:
> @deads2k Can you explain what you mean by "...take the HttpHeaders support as a follow on item... without plumbing them through.".

The `HTTPHeaders` field is used by docker to add opaque headers to client requests.  The new format supports this field, but actually implementing in the kubelet will take additional work.  I'd prefer to avoid having that feature block this pull.  This pull is very useful even if the `HTTPHeaders` field doesn't work as expected.

deads2k commented:
> What do you think about this approach? I will add more tests in pkg/api/validation/validation_test.go later.

Looking at how the code has fallen out, I might have factored it as a method that takes some bytes in and attempts to produce a `DockerConfig` from them.  Something like:

```
func DockerConfigFrom(bytes []bytes) (DockerConfig, error){
    // first try to parse as the new format
    // if that fails, try to parse the old format
    // if the old format parse fails, return the error from the new format parse
}
```

That eliminates the ability to make use of the `HTTPHeaders` field, but it would keep the paths unified and obviates the need for a new type of secret (yeah, lavalamp called that one).  You'd also be able to build a list of the exact files you wanted to check in `ReadDockerConfigFile` and have a single block of code for testing, reading, and parsing them.

bgrant0607 commented:
cc @liggitt @kubernetes/goog-node

erictune commented:
I'm on vacation this week (:house: :mouse:) and will take a look next week.

ngbinh commented:
ping

erictune commented:
@dchen1107 I think we might want to take this PR, maybe for v1.1.
Any concerns?  Also, is there anyone on the hosted side that we want to look at this?

dchen1107 commented:
Ahh, this is dup of #13894. cc/ @mattmoor

erictune commented:
LGTM
